### PR TITLE
Performance improvements for ConditionalDistribution

### PR DESCRIPTION
**Pull request by dbarbier**
Call computePDF(Sample) instead of looping over computePDF(Point),
this is much faster.

### CODE REVIEW

**Comment by dbarbier** [comment]
`t_PosteriorDistribution_std` runs 30% faster on my machine with this commit (15s instead of 22s), I guess that speed-up on computeExpectation itself is quite important.
It has not been fully tested though, this is almost a proof-of-concept.

File: lib/src/Uncertainty/Distribution/ConditionalDistribution.cxx [hunk 1]
```diff
@@ -528,33 +553,42 @@ Point ConditionalDistribution::computeExpectation(const Function & f,
...
     // Skip the current integration point if the current sub-manifold is outside of the integration region
-    if (rejectNode) continue;
-    // Now, complete by the discretization of the continuous part using Gauss integration
-    for (UnsignedInteger j = 0; j < continuousAtomsNumber; ++j)
+    if (!rejectNode)
     {
-      // Complete the filling of theta using the Gauss integration node
-      const Point continuousNode(continuousNodes_[j]);
-      for (UnsignedInteger k = 0; k < continuousDimension; ++k)
-        currentTheta[continuousMarginalsIndices_[k]] = continuousLowerBounds_[k] + 0.5 * (1.0 + continuousNode[k]) * (subPoint[k] - continuousLowerBounds_[k]);
```
Comment (dbarbier) [inline lib/src/Uncertainty/Distribution/ConditionalDistribution.cxx]: @regislebrun I rewrite this portion of code to perform the same computations as before (if there are changes, it is a bug), but I find it very suspicious.  I would expect to reset currentTheta on its value before loop on j, something like
```
 const Point discreteTheta(currentTheta);
 for (UnsignedInteger j = 0; j < continuousAtomsNumber; ++j)
 {
   const Point continuousNode(continuousNodes_[j]);
   for (UnsignedInteger k = 0; k < continuousDimension; ++k)
     discreteTheta[continuousMarginalsIndices_[k]] = continuousLowerBounds_[k] + 0.5 * (1.0 + continuousNode[k]) * (subPoint[k] - continuousLowerBounds_[k]);
     const Scalar w = conditioningDistribution_.computePDF(discreteTheta) * continuousWeights_[j];
   ...
 }
```

Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/ConditionalDistribution.cxx]: @dbarbier Ok I will check this tomorrow. The Bayesian part of the distributions (ConditionalDistribution, BayesDistribution and PosteriorDistribution) is at an early stage of validation, and should be refactored. The main lines of the algorithms are ok in my view, but there is still a lot to do to get something efficient and robust. Thanks for the effort.

Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/ConditionalDistribution.cxx]: @dbarbier The ConditionalDistribution::computeConditionalDistribution() method is strange as a whole. To be honest, I don't even remember its precise role, the comment being much too laconic. A good thing would be to allow the user to choose what kind of WeightedExperiment she want to use for the discretization. The default GaussProductExperiment is good only when the conditioned distribution depends smoothly on its parameters, which is not always the case (eg: a Uniform with random bounds). In this case, a LowDiscrepancyExperiment would do a better job.

Comment (sofianehaddad) [reply lib/src/Uncertainty/Distribution/ConditionalDistribution.cxx]: @regislebrun yes but in an other PR. The objective here is to improve only performance and this improvement is required whatever the used WeightedExperiment

**Comment by sofianehaddad** [comment]
@dbarbier could you please rebase ?

**Comment by dbarbier** [comment]
Right, my point is that we should write code with performance in mind.  We all know that Sample indexing is convenient but may have poor performance, thus we should always favor `Sample(i,j)` over `Sample[i][j]`.  And likewise it is always faster to compute a function on a Sample rather than evaluating it for each Point.

**Comment by regislebrun** [comment]
@dbarbier It looks like it would be a good idea to replace each occurrence of sample[i][j] by sample(i,j) in the library. A quick count using grep -rl '\]\[' lib/src | wc -l gives a number of 168 files to change, which is large. I can do it but I suspect it will be a pain to review then to merge. Do you have an advice on how to manage this kind of change?

**Comment by regislebrun** [comment]
otherwise, +1 for the PR

**Pull request merged**